### ISSUE TITLE
State machine compilation failed

### ISSUE THREAD

**Issue by MatteoSchiavon** [OpenModelica/OpenModelica#11642]
### Description

I'm trying to test the clocked state machine "Operate" in a "TestOperate" model. The SM takes only one clocked input and it has 2 states: run and freeze. When the input goes below 0.5 the transition from run to freeze occurs and an inner counter gets increased. The clocked variable "sample1.y" is initialized to 1, which is the starting value of the continuous step source signal that will eventually ramp to 0.

Below there are the model implementations without annotation of the "TestOperate.mo", but I put a screenshot of it to make it clear. 

### Steps to Reproduce

1. run like they are implemented below:
2. run but uncomment the lines:
    //outer output Integer count;
  //equation
    //count = 0;

............................................................................
**model Operate**
  
  Modelica.Blocks.Interfaces.RealInput u annotation(Placement(transformation(origin = {-108, 0}, extent = {{-20, -20}, {20, 20}}), transformation(origin = {-100, 0}, extent = {{-20, -20}, {20, 20}})));
  inner Integer count(start = 0);

  block Run
    //outer output Integer count;
  //equation
    //count = 0;
  end Run;
  Run run;

  block Freeze
    outer output Integer count;
  equation
    count = previous(count) + 1;
  end Freeze;
  Freeze freeze;
  
equation
  initialState(run);
  transition(run, freeze, u < 0.5, immediate = false);
  
**end Operate;**
............................................................................
**model TestOperate**

  Modelica.Clocked.RealSignals.Sampler.SampleClocked sample1(y.start=1);
  Modelica.Clocked.ClockSignals.Clocks.PeriodicExactClock periodicClock1(factor = 10, resolution = Modelica.Clocked.Types.Resolution.ms);
  Modelica.Blocks.Sources.Ramp ramp(height = -1, duration = 1, offset = 1, startTime = 1);
  Operate operate;

equation
  connect(periodicClock1.y, sample1.clock);
  connect(ramp.y, sample1.u);
  connect(operate.u, sample1.y);

**end TestOperate;**
............................................................................

### Expected Behavior

Query 1. I am not sure about the origin of the error "Compilation process failed" but it deals with "mingw32-make".
Query 2. The error becomes "Too many equations, over-determined system. The model has 2 equation(s) and 1 variable(s)." I get the meaning of this but, from the Langauge Specification I would expect the compiler to take care of the system determination since they say "State machines have the important feature that at one clock tick, there is only one assignment to every variable".

In general, I can't find a way to make sure that my state machine is clocked so, could that be the main issue? I am expecting the clock to be inferred by the input signal but that's all I can think of.

### Screenshots

![image](https://github.com/OpenModelica/OpenModelica/assets/142384341/e550d658-056f-4a7b-8e86-be7ccecf4faa)

### Version and OS

 - OpenModelica: v1.23.0-dev-28-gd1144cfa88 (64-bit)
 - OMSimulator: v2.1.1.post236-g634b019-mingw-notlm-debug
 - OS Windows 10, 64 bit
 - Versions of used Modelica libraries if applicable: 4.0.0

**Comment by casella** [comment]
@MatteoSchiavon the state machines are still not (fully) implemented in the new backend, which is used by default. Maybe @perost can comment on that?

**Comment by MatteoSchiavon** [comment]
Hello,
Do you have a time estimation of how long it'll take to develop the state machines in OpenModelica?
Best regards,
Matteo

**Comment by perost** [comment]
> We still don't have precise plans for this. I think @perost will have some time in the next few months, if this is useful for your work we can consider that, we have to do anyway at some point in time.
> 
> @perost?

I don't have much to add to that, I don't know what would be involved in implementing state machine support so I can't really say how much time it would take.

**Comment by casella** [comment]
Don't we have it in the old frontend?

**Comment by perost** [comment]
> Don't we have it in the old frontend? Maybe I was mistaken about that, I though it was just a matter of porting it from OF to NF

Yes, but I'm not sure how complete it is or how much work is needed to port it. There's one file containing a couple of thousand lines of code that does something with the instantiated state machines using the old OF structures, so that might need to be completely rewritten to use the NF structures. And I guess there's also more things spread out in the compiler too.

> Regarding implementation, the state machine semantics is specified in Modelica in Section [17.3.4](https://specification.modelica.org/maint/3.6/state-machines.html#semantics-summary) of the MLS. One option is to actually use this specification literally, i.e. generating the corresponding Modelica code. Maybe it's not the most efficient implementation, but it should work.
> 
> @perost what do you think?

The current implementation transforms the state machines to "data-flow equations", whatever that is. Maybe that's not really necessary and we could use a simpler implementation, I'm not really sure.

**Comment by casella** [comment]
Do you know who wrote the implementation in the OF?

**Comment by perost** [comment]
> Do you know who wrote the implementation in the OF?

https://github.com/OpenModelica/OpenModelica/commit/c30ebadd3d29e0071beac301baf8db135701a050

**Comment by casella** [comment]
OK.

@bernhard-thiele would you be available for a webmeeting where you explain to Per how did you implement the state machines in the old frontend? We need to figure out if it's feasible to port your implementation to the new one, or rather do it from scratch. 

Thanks!

**Comment by rcerny** [comment]
I have the same issue, and I am wondering if there is a work-around in the meanwhile? 
Many Thanks for update!
